quic-rpc-derive/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+                }
+                impl ::quic_rpc::pattern::client_streaming::ClientStreamingMsg<#service_name> for #request_type {
+                    type Update = #update;
+                    type Response = #response;
+                }
+            }
+        }
+        TRY_SERVER_STREAMING => {
+            let create_error = args.get("create_error", pat, attr_span)?;
+            let item_error = args.get("item_error", pat, attr_span)?;
+            let item = args.get("item", pat, attr_span)?;
+            quote! {
+                impl ::quic_rpc::message::Msg<#service_name> for #request_type {
+                    type Pattern = ::quic_rpc::pattern::try_server_streaming::TryServerStreaming;
+                }
+                impl ::quic_rpc::pattern::try_server_streaming::TryServerStreamingMsg<#service_name> for #request_type {
+                    type CreateError = #create_error;
+                    type ItemError = #item_error;
+                    type Item = #item;
+                }
+            }
+        }
+        _ => return Err(syn::Error::new(attr_span, "Unknown RPC pattern")),
+    };
+    args.check_empty(attr_span)?;
+
+    Ok(res)
+}
+
+#[proc_macro_attribute]
+pub fn rpc_requests(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let mut input = parse_macro_input!(item as DeriveInput);
+    let service_name = parse_macro_input!(attr as Ident);
+
+    let data_enum = match &mut input.data {
+        Data::Enum(data_enum) => data_enum,
+        _ => {
+            return syn::Error::new(input.span(), "RpcRequests can only be applied to enums")
+                .to_compile_error()
+                .into()
+        }
+    };
+
+    let mut additional_items = Vec::new();
+    let mut types = HashSet::new();
+
+    for variant in &mut data_enum.variants {
+        // Check field structure for every variant
+        let request_type = match &variant.fields {
+            Fields::Unnamed(fields) if fields.unnamed.len() == 1 => &fields.unnamed[0].ty,
+            _ => {
+                return syn::Error::new(
+                    variant.span(),
+                    "Each variant must have exactly one unnamed field",
+                )
+                .to_compile_error()
+                .into()
+            }
+        };
+
+        if !types.insert(request_type.to_token_stream().to_string()) {
+            return syn::Error::new(input.span(), "Each variant must have a unique request type")
+                .to_compile_error()
+                .into();
+        }
+
+        //
+        let mut rpc_attr = Vec::new();
+        variant.attrs.retain(|attr| {
+            for ident in IDENTS {
+                if attr.path.is_ident(ident) {
+                    rpc_attr.push((ident, attr.clone()));
+                    return false;
+                }
+            }
+            true
+        });
+
+        // Fail if there are multiple RPC patterns
+        if rpc_attr.len() > 1 {
```

Comment:
what about if this is empty?

Reply:
I don't mind empty. E.g. you got one case for which you want to define the traits manually for whatever reason. The macro supports only a limited set of interaction patterns, it is not open.